### ISSUE TITLE
StatefulSets are orphaned on deletion of cassandradatacenter

### ISSUE THREAD

**Issue by rltvty** [datastax/cass-operator#178]
I'm seeing an issue when I delete my Cassandradatacenter config, it destroys the first statefulset but not the remaining ones.

I'm running size 3 rack-aware cluster in AWS EKS, defined as follows:

```
apiVersion: cassandra.datastax.com/v1beta1
kind: CassandraDatacenter
metadata:
  name: marlin
  namespace: sandbox
spec:
  clusterName: marlin-db
  config:
    cassandra-yaml:
      authenticator: org.apache.cassandra.auth.PasswordAuthenticator
      authorizer: org.apache.cassandra.auth.CassandraAuthorizer
      num_tokens: 8
      role_manager: org.apache.cassandra.auth.CassandraRoleManager
    jvm-options:
      additional-jvm-opts:
      - -Dcom.sun.management.jmxremote
      - -Dcom.sun.management.jmxremote.authenticate=false
      - -Dcom.sun.management.jmxremote.ssl=false
      - -Dcom.sun.management.jmxremote.local.only=false
      - -Dcom.sun.management.jmxremote.port=7199
      - -Dcom.sun.management.jmxremote.rmi.port=7199
      heap_size_young_generation: 2G
      initial_heap_size: 1G
      max_heap_size: 4G
  managementApiAuth:
    insecure: {}
  racks:
  - name: us-west-2a
  - name: us-west-2b
  - name: us-west-2c
  resources:
    limits:
      cpu: 1000m
      memory: 4Gi
    requests:
      cpu: 500m
      memory: 2Gi
  serverType: cassandra
  serverVersion: 3.11.6
  size: 3
  storageConfig:
    cassandraDataVolumeClaimSpec:
      accessModes:
      - ReadWriteOnce
      resources:
        requests:
          storage: 5Gi
      storageClassName: gp2
```

Logs from the operator pod report the following:

```
{"level":"info","ts":1595989198.4807413,"logger":"reconciliation_handler","msg":"======== handler::Reconcile has been called","requestNamespace":"sandbox","requestName":"marlin","loopID":"b8155cae-fb56-425d-b4f1-9827704a4c78"}
{"level":"info","ts":1595989198.4807637,"logger":"reconciliation_handler","msg":"handler::CreateReconciliationContext","requestNamespace":"sandbox","requestName":"marlin","loopID":"b8155cae-fb56-425d-b4f1-9827704a4c78","namespace":"sandbox"}
{"level":"info","ts":1595989198.480812,"logger":"reconciliation_handler","msg":"handler::calculateReconciliationActions","requestNamespace":"sandbox","requestName":"marlin","loopID":"b8155cae-fb56-425d-b4f1-9827704a4c78","namespace":"sandbox","datacenterName":"marlin","clusterName":"marlin-db"}
{"level":"info","ts":1595989198.4903657,"logger":"reconciliation_handler","msg":"reconciler::deletePVCs","requestNamespace":"sandbox","requestName":"marlin","loopID":"b8155cae-fb56-425d-b4f1-9827704a4c78","namespace":"sandbox","datacenterName":"marlin","clusterName":"marlin-db"}
{"level":"info","ts":1595989198.4903955,"logger":"reconciliation_handler","msg":"reconciler::listPVCs","requestNamespace":"sandbox","requestName":"marlin","loopID":"b8155cae-fb56-425d-b4f1-9827704a4c78","namespace":"sandbox","datacenterName":"marlin","clusterName":"marlin-db"}
{"level":"info","ts":1595989198.490468,"logger":"reconciliation_handler","msg":"Found PVCs for cassandraDatacenter","requestNamespace":"sandbox","requestName":"marlin","loopID":"b8155cae-fb56-425d-b4f1-9827704a4c78","namespace":"sandbox","datacenterName":"marlin","clusterName":"marlin-db","cassandraDatacenterNamespace":"sandbox","cassandraDatacenterName":"marlin","numPVCs":3}
{"level":"info","ts":1595989198.4972866,"logger":"reconciliation_handler","msg":"Deleted PVC","requestNamespace":"sandbox","requestName":"marlin","loopID":"b8155cae-fb56-425d-b4f1-9827704a4c78","namespace":"sandbox","datacenterName":"marlin","clusterName":"marlin-db","cassandraDatacenterNamespace":"sandbox","cassandraDatacenterName":"marlin","pvcNamespace":"sandbox","pvcName":"server-data-marlin-db-marlin-us-west-2a-sts-0"}
{"level":"info","ts":1595989198.5036826,"logger":"reconciliation_handler","msg":"Deleted PVC","requestNamespace":"sandbox","requestName":"marlin","loopID":"b8155cae-fb56-425d-b4f1-9827704a4c78","namespace":"sandbox","datacenterName":"marlin","clusterName":"marlin-db","cassandraDatacenterNamespace":"sandbox","cassandraDatacenterName":"marlin","pvcNamespace":"sandbox","pvcName":"server-data-marlin-db-marlin-us-west-2b-sts-0"}
{"level":"info","ts":1595989198.5101688,"logger":"reconciliation_handler","msg":"Deleted PVC","requestNamespace":"sandbox","requestName":"marlin","loopID":"b8155cae-fb56-425d-b4f1-9827704a4c78","namespace":"sandbox","datacenterName":"marlin","clusterName":"marlin-db","cassandraDatacenterNamespace":"sandbox","cassandraDatacenterName":"marlin","pvcNamespace":"sandbox","pvcName":"server-data-marlin-db-marlin-us-west-2c-sts-0"}
2020/07/29 02:19:58 http: TLS handshake error from 10.110.59.250:59228: remote error: tls: bad certificate
{"level":"info","ts":1595989198.5372877,"logger":"api","msg":"Validating webhook called for update"}
{"level":"info","ts":1595989198.5438676,"logger":"api","msg":"Validating webhook called for update"}
2020/07/29 02:19:58 http: TLS handshake error from 10.110.59.250:59232: remote error: tls: bad certificate
{"level":"info","ts":1595989198.5640762,"logger":"reconciliation_handler","msg":"Reconcile loop completed","requestNamespace":"sandbox","requestName":"marlin","loopID":"b8155cae-fb56-425d-b4f1-9827704a4c78","duration":0.083447013}
{"level":"info","ts":1595989198.5642374,"logger":"reconciliation_handler","msg":"======== handler::Reconcile has been called","requestNamespace":"sandbox","requestName":"marlin","loopID":"b7b42f2d-9c44-4ffa-a4c9-fc9a16fb092e"}
{"level":"info","ts":1595989198.5642517,"logger":"reconciliation_handler","msg":"handler::CreateReconciliationContext","requestNamespace":"sandbox","requestName":"marlin","loopID":"b7b42f2d-9c44-4ffa-a4c9-fc9a16fb092e","namespace":"sandbox"}
{"level":"error","ts":1595989198.564267,"logger":"reconciliation_handler","msg":"error in retrieveDatacenter","requestNamespace":"sandbox","requestName":"marlin","loopID":"b7b42f2d-9c44-4ffa-a4c9-fc9a16fb092e","namespace":"sandbox","error":"CassandraDatacenter.cassandra.datastax.com \"marlin\" not found","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/pkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128\ngithub.com/datastax/cass-operator/operator/pkg/reconciliation.CreateReconciliationContext\n\t/cass-operator/operator/pkg/reconciliation/context.go:69\ngithub.com/datastax/cass-operator/operator/pkg/reconciliation.(*ReconcileCassandraDatacenter).Reconcile\n\t/cass-operator/operator/pkg/reconciliation/handler.go:102\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:256\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:232\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:211\nk8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/wait/wait.go:152\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/wait/wait.go:153\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/wait/wait.go:88"}
{"level":"info","ts":1595989198.5643423,"logger":"reconciliation_handler","msg":"CassandraDatacenter resource not found. Ignoring since object must be deleted.","requestNamespace":"sandbox","requestName":"marlin","loopID":"b7b42f2d-9c44-4ffa-a4c9-fc9a16fb092e"}
{"level":"info","ts":1595989198.5643475,"logger":"reconciliation_handler","msg":"Reconcile loop completed","requestNamespace":"sandbox","requestName":"marlin","loopID":"b7b42f2d-9c44-4ffa-a4c9-fc9a16fb092e","duration":0.000225703}
{"level":"info","ts":1595989198.5643744,"logger":"reconciliation_handler","msg":"======== handler::Reconcile has been called","requestNamespace":"sandbox","requestName":"marlin","loopID":"14a1ffcd-69bd-457d-b2a9-6619bf37d58c"}
{"level":"info","ts":1595989198.5643816,"logger":"reconciliation_handler","msg":"handler::CreateReconciliationContext","requestNamespace":"sandbox","requestName":"marlin","loopID":"14a1ffcd-69bd-457d-b2a9-6619bf37d58c","namespace":"sandbox"}
{"level":"error","ts":1595989198.564391,"logger":"reconciliation_handler","msg":"error in retrieveDatacenter","requestNamespace":"sandbox","requestName":"marlin","loopID":"14a1ffcd-69bd-457d-b2a9-6619bf37d58c","namespace":"sandbox","error":"CassandraDatacenter.cassandra.datastax.com \"marlin\" not found","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/pkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128\ngithub.com/datastax/cass-operator/operator/pkg/reconciliation.CreateReconciliationContext\n\t/cass-operator/operator/pkg/reconciliation/context.go:69\ngithub.com/datastax/cass-operator/operator/pkg/reconciliation.(*ReconcileCassandraDatacenter).Reconcile\n\t/cass-operator/operator/pkg/reconciliation/handler.go:102\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:256\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:232\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:211\nk8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/wait/wait.go:152\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/wait/wait.go:153\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/wait/wait.go:88"}
{"level":"info","ts":1595989198.5644274,"logger":"reconciliation_handler","msg":"CassandraDatacenter resource not found. Ignoring since object must be deleted.","requestNamespace":"sandbox","requestName":"marlin","loopID":"14a1ffcd-69bd-457d-b2a9-6619bf37d58c"}
{"level":"info","ts":1595989198.5644314,"logger":"reconciliation_handler","msg":"Reconcile loop completed","requestNamespace":"sandbox","requestName":"marlin","loopID":"14a1ffcd-69bd-457d-b2a9-6619bf37d58c","duration":0.000067277}
{"level":"info","ts":1595989199.6777594,"logger":"reconciliation_handler","msg":"======== handler::Reconcile has been called","requestNamespace":"sandbox","requestName":"marlin","loopID":"e77becf3-bacb-40eb-9ebe-515042769db9"}
{"level":"info","ts":1595989199.6779435,"logger":"reconciliation_handler","msg":"handler::CreateReconciliationContext","requestNamespace":"sandbox","requestName":"marlin","loopID":"e77becf3-bacb-40eb-9ebe-515042769db9","namespace":"sandbox"}
{"level":"error","ts":1595989199.6783075,"logger":"reconciliation_handler","msg":"error in retrieveDatacenter","requestNamespace":"sandbox","requestName":"marlin","loopID":"e77becf3-bacb-40eb-9ebe-515042769db9","namespace":"sandbox","error":"CassandraDatacenter.cassandra.datastax.com \"marlin\" not found","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/pkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128\ngithub.com/datastax/cass-operator/operator/pkg/reconciliation.CreateReconciliationContext\n\t/cass-operator/operator/pkg/reconciliation/context.go:69\ngithub.com/datastax/cass-operator/operator/pkg/reconciliation.(*ReconcileCassandraDatacenter).Reconcile\n\t/cass-operator/operator/pkg/reconciliation/handler.go:102\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:256\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:232\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:211\nk8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/wait/wait.go:152\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/wait/wait.go:153\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/wait/wait.go:88"}
{"level":"info","ts":1595989199.6785257,"logger":"reconciliation_handler","msg":"CassandraDatacenter resource not found. Ignoring since object must be deleted.","requestNamespace":"sandbox","requestName":"marlin","loopID":"e77becf3-bacb-40eb-9ebe-515042769db9"}
{"level":"info","ts":1595989199.678664,"logger":"reconciliation_handler","msg":"Reconcile loop completed","requestNamespace":"sandbox","requestName":"marlin","loopID":"e77becf3-bacb-40eb-9ebe-515042769db9","duration":0.000929498}
{"level":"info","ts":1595989199.6850798,"logger":"reconciliation_handler","msg":"======== handler::Reconcile has been called","requestNamespace":"sandbox","requestName":"marlin","loopID":"abafa661-e6fb-4b31-84c2-c901a9eeb6c0"}
{"level":"info","ts":1595989199.6853395,"logger":"reconciliation_handler","msg":"handler::CreateReconciliationContext","requestNamespace":"sandbox","requestName":"marlin","loopID":"abafa661-e6fb-4b31-84c2-c901a9eeb6c0","namespace":"sandbox"}
{"level":"error","ts":1595989199.6858416,"logger":"reconciliation_handler","msg":"error in retrieveDatacenter","requestNamespace":"sandbox","requestName":"marlin","loopID":"abafa661-e6fb-4b31-84c2-c901a9eeb6c0","namespace":"sandbox","error":"CassandraDatacenter.cassandra.datastax.com \"marlin\" not found","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/pkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128\ngithub.com/datastax/cass-operator/operator/pkg/reconciliation.CreateReconciliationContext\n\t/cass-operator/operator/pkg/reconciliation/context.go:69\ngithub.com/datastax/cass-operator/operator/pkg/reconciliation.(*ReconcileCassandraDatacenter).Reconcile\n\t/cass-operator/operator/pkg/reconciliation/handler.go:102\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:256\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:232\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:211\nk8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/wait/wait.go:152\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/wait/wait.go:153\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/wait/wait.go:88"}
{"level":"info","ts":1595989199.6863525,"logger":"reconciliation_handler","msg":"CassandraDatacenter resource not found. Ignoring since object must be deleted.","requestNamespace":"sandbox","requestName":"marlin","loopID":"abafa661-e6fb-4b31-84c2-c901a9eeb6c0"}
{"level":"info","ts":1595989199.6865418,"logger":"reconciliation_handler","msg":"Reconcile loop completed","requestNamespace":"sandbox","requestName":"marlin","loopID":"abafa661-e6fb-4b31-84c2-c901a9eeb6c0","duration":0.001478445}
{"level":"info","ts":1595989199.6867383,"logger":"reconciliation_handler","msg":"======== handler::Reconcile has been called","requestNamespace":"sandbox","requestName":"marlin","loopID":"9e9e52ad-39cf-4f73-b50e-b1e725d67fa1"}
{"level":"info","ts":1595989199.6869369,"logger":"reconciliation_handler","msg":"handler::CreateReconciliationContext","requestNamespace":"sandbox","requestName":"marlin","loopID":"9e9e52ad-39cf-4f73-b50e-b1e725d67fa1","namespace":"sandbox"}
{"level":"error","ts":1595989199.6871238,"logger":"reconciliation_handler","msg":"error in retrieveDatacenter","requestNamespace":"sandbox","requestName":"marlin","loopID":"9e9e52ad-39cf-4f73-b50e-b1e725d67fa1","namespace":"sandbox","error":"CassandraDatacenter.cassandra.datastax.com \"marlin\" not found","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/pkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128\ngithub.com/datastax/cass-operator/operator/pkg/reconciliation.CreateReconciliationContext\n\t/cass-operator/operator/pkg/reconciliation/context.go:69\ngithub.com/datastax/cass-operator/operator/pkg/reconciliation.(*ReconcileCassandraDatacenter).Reconcile\n\t/cass-operator/operator/pkg/reconciliation/handler.go:102\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:256\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:232\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:211\nk8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/wait/wait.go:152\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/wait/wait.go:153\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/wait/wait.go:88"}
{"level":"info","ts":1595989199.6872232,"logger":"reconciliation_handler","msg":"CassandraDatacenter resource not found. Ignoring since object must be deleted.","requestNamespace":"sandbox","requestName":"marlin","loopID":"9e9e52ad-39cf-4f73-b50e-b1e725d67fa1"}
{"level":"info","ts":1595989199.6872733,"logger":"reconciliation_handler","msg":"Reconcile loop completed","requestNamespace":"sandbox","requestName":"marlin","loopID":"9e9e52ad-39cf-4f73-b50e-b1e725d67fa1","duration":0.000555763}
```

**Comment by rltvty** [comment]
these are the statefulsets that remain after the deletion: 

```
marlin-db-marlin-us-west-2b-sts   1/1     74m
marlin-db-marlin-us-west-2c-sts   1/1     74m
```

**Comment by rltvty** [comment]
perhaps this is fixed by #121 ? Is there a build published with this fix that I can try?

**Comment by jimdickinson** [comment]
Yes - which operator version are you using?

**Comment by jimdickinson** [comment]
Thank you for the really detailed report with your full cassdc configuration - could you add on any information for how it came to be this way? Did you just create? Create and update N times? The bug that #121 fixed would've always effected all statefulsets is why I ask.

**Comment by rltvty** [comment]
Ah, I thought I was already using 1.3, but I'm still on 1.2.  I'll try upgrading today.

This happens when I create a cluster, and let it come up to steady-state, and then delete it.  Just a single create and then delete about 30 mins later, no updates in-between.

I'll let you know if 1.3 helps.  Thanks!

**Comment by rltvty** [comment]
This is fixed in 1.3!  Sorry for using an old version.

**Comment by jimdickinson** [comment]
Awesome! I did add a test to make sure this didn't regress, so good news all around!
